File: src/editor/filter/leaf/options/multi/CardTypeFilter.java
package editor.filter.leaf.options.multi;

import java.util.HashSet;

import com.google.gson.JsonElement;
import com.google.gson.JsonPrimitive;

import editor.database.card.Card;
import editor.filter.Filter;
import editor.filter.FilterAttribute;

/**
 * This class represents a filter that groups cards by card type.
 *
 * @author Alec Roelke
 */
public class CardTypeFilter extends MultiOptionsFilter<String>
{
    /**
     * List of all types that appear on cards (including ones that appear on Unglued and Unhinged cards, whose
     * type lines were not updated for the most modern templating).
     */
    public static String[] typeList = {};

    /**
     * Create a new CardTypeFilter.
     */
    public CardTypeFilter()
    {
        super(FilterAttribute.CARD_TYPE, Card::types);
    }

    @Override
    protected String convertFromString(String str)
    {
        return str;
    }

    @Override
    public Filter copy()
    {
        CardTypeFilter filter = (CardTypeFilter)FilterAttribute.createFilter(FilterAttribute.CARD_TYPE);
        filter.contain = contain;
        filter.selected = new HashSet<>(selected);
        return filter;
    }

    @Override
    protected JsonElement convertToJson(String item)
    {
        return new JsonPrimitive(item);
    }

    @Override
    protected String convertFromJson(JsonElement item)
    {
        return item.getAsString();
    }
}

File: src/main/scala/editor/filter/leaf/options/multi/CardTypeFilter.scala
package editor.filter.leaf.options.multi

import editor.database.attributes.CardAttribute
import editor.filter.FaceSearchOptions
import editor.util.Containment

/**
 * Object containing global information about card types.
 * @author Alec Roelke
 */
object CardTypeFilter {
  /** List of all card types. */
  var typeList = Seq[String]()
}

/**
 * Filter that groups cards based on card types.
 * @author Alec Roelke
 */
final case class CardTypeFilter(faces: FaceSearchOptions = FaceSearchOptions.ANY, contain: Containment = Containment.AnyOf, selected: Set[String] = Set.empty) extends MultiOptionsFilter[String, CardTypeFilter] {
  override def attribute = CardAttribute.CardType
  override val unified = false
  override val values = _.types
}